### DISCUSSION TITLE
Table in /farming is empty

### DISCUSSION THREAD

**Discussion by elexx** [guydavis/machinaris#123] [Q&A] [answered]
**Describe the bug**
When opening http://myfullnode/farming the plot-table is empty, although I already plotted some plots:
![image](https://user-images.githubusercontent.com/221837/123557240-1a36e600-d790-11eb-917f-226d529321d8.png)

**To Reproduce**
1. docker pull "ghcr.io/guydavis/machinaris:test"
2. docker run [...nothing special here, just provide the tmp-volumes and a final dst]
3. inital setup (wallet, pk)
4. switch config to madmax, start plotter
5. see plots arriving at landing-page, and see "expected time to win" shrink
6. navigate to /farming
7. see empty table, like in the screenshot above

**Expected behavior**
A filled table like in https://github.com/guydavis/machinaris#farming-view

**System setup:**
 - OS: Docker
 - Docker version 20.10.7, build f0df35096d
 - Kernel Version: 5.9.16-1-MANJARO
 - Operating System: Manjaro Linux
 - Architecture: x86_64
 - CPUs: 24
 - Total Memory: 125.8GiB
 - Machinaris branch: test
 - Machinaris version number: 0.4.0, but I'm on test, so yea.

**Config**
I'm sorry, but I'm not sure what config I should paste here.

**Additional context & screenshots**
I did a bit of digging and found that the sqlite database does not contain any plots-rows:
```
sqlite3 machinaris.db
SQLite version 3.35.5 2021-04-19 18:32:05
Enter ".help" for usage hints.
sqlite> SELECT * FROM plots;
sqlite> SELECT COUNT(*) FROM plots;
0
sqlite> .quit
```
Which explains the empty table in my UI. However I did not dig further. Where should the table be filled? Can I provide any additional information?

Thanks

**Comment by guydavis** [comment]
Good day, please provide the exact and entire Docker run command or compose file you are using to run Machinaris.  Thanks!

**Comment by elexx** [comment]
I started the container with:
```
docker run \
 --name="machinaris" \
 -d \
 -t \
 -h $(hostname -s) \
 -p 8444:8444 \
 -p 8926:8926 \
 -p 8927:8927 \
 -e TZ="Europe/Vienna" \
 -v "/home/$(whoami)/.machinaris":"/root/.chia":"rw" \
 -v "/plotroot":"/plotting":"rw" \
 -v "/run/media/$(whoami)":"/plots":"rw" \
 --mount "type=tmpfs,tmpfs-size=110G,destination=/plotting_ram" \
 "ghcr.io/guydavis/machinaris:test"

**Comment by bland328** [comment]
I'm also seeing this after updating to Machinaris 0.4.0

I've changed nothing about the way the container is started _(on an Unraid server, FWIW)_, and the result is that Machinaris 0.4.0 claims to be farming the correct number of plots, but lists no plots under the Farming tab.

_Edit: I also checked my sqlite3 db, but_ did _find the correct number of records, and then subsequently found that Machinaris had started listing my plots. Perhaps patience was called for in my case?_

**Comment by guydavis** [comment]
@elexx Thanks for the run command.  I read that your plots are all held in a single mount point on your host OS 
 `/run/media/$(whoami)` with no subfolders.

If that is true, then it would be good to see the Machinaris logs.  Please attach the last 100 lines or so of these files immediately after you browse to the Farming page.
- /home/$(whoami)/.machinaris/machinaris/logs/*.log

If those logs, don't make the issue clear, you can enter debug mode (more logging):
```
docker exec -it machinaris bash
export FLASK_ENV=development
/machinaris/scripts/start_machinaris.sh
tail -f /root/.chia/machinaris/logs/apisrv.log
```

![image](https://user-images.githubusercontent.com/12936/123653840-bdb4e480-d7ea-11eb-8008-7b30db8824b8.png)

**Comment by elexx** [comment]
`/run/media/$(whoami)` is my base-path for my external HDDs, so there are subfolders (mountpoints) under it:
```
$ df -h | grep /run/media
/dev/sda1       [...] /run/media/admin/c2432ac8-ce40-458a-98f7-195291099312
/dev/sdb1       [...] /run/media/admin/3ec8901d-d783-48ec-834c-b7a9e84f2f39
/dev/sdg1       [...] /run/media/admin/5cd3212e-c13f-4883-952f-86e056d4b86d
/dev/sdc1       [...] /run/media/admin/6d9f0919-17dc-4f42-864a-6aa51260427d
/dev/sdd1       [...] /run/media/admin/be46654e-ed23-4c7f-8a16-7413fa822969
/dev/sde1       [...] /run/media/admin/1ae119fa-294e-4449-b28f-a27141b9a77c
/dev/sdf1       [...] /run/media/admin/9ea8907d-f34c-4114-8c02-6e83e8e04651
/dev/sdh1       [...] /run/media/admin/f5d4dd23-1035-4115-b4ca-30e6a3a1d1f5
/dev/sdi1       [...] /run/media/admin/19c0dbe2-30e4-4734-92d9-2a81b049111b
/dev/sdj1       [...] /run/media/admin/9f78851b-437f-4375-8791-56c5d0f60ddb
```
Opening `/farming` does not produce any logs at all. The filesize of all 3 logfiles (apisrv.log, migration.log, webui.log) stay the same (before opening /farming vs after opening /farming).

----
```
root@fullnode:/chia-blockchain# export FLASK_ENV=development
root@fullnode:/chia-blockchain# /machinaris/scripts/start_machinaris.sh
Configuring Chia...
Configuring Plotman...
Configuring Chiadog...
Starting Chiadog...
Starting Machinaris API server...
Starting Machinaris Web server...
Completed startup.  Browse to port 8926.
root@fullnode:/chia-blockchain# tail -f /root/.chia/machinaris/logs/apisrv.log
2021-06-28 16:46:31,301 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-06-28 16:46:31,302 INFO sqlalchemy.engine.Engine SELECT workers.hostname AS workers_hostname, workers.mode AS workers_mode, workers.services AS workers_services, workers.url AS workers_url, workers.config AS workers_config, workers.latest_ping_result AS workers_latest_ping_result, workers.ping_success_at AS workers_ping_success_at, workers.created_at AS workers_created_at, workers.updated_at AS workers_updated_at 
FROM workers 
WHERE workers.hostname = ?
2021-06-28 16:46:31,302 INFO sqlalchemy.engine.Engine [generated in 0.00016s] ('192.168.10.105',)
2021-06-28 16:46:31,303 INFO sqlalchemy.engine.Engine ROLLBACK
[2021-06-28 16:46:23 +0200] [720] [INFO] Handling signal: term
[2021-06-28 16:46:23 +0200] [989] [INFO] Worker exiting (pid: 989)
[2021-06-28 16:46:23 +0200] [1014] [INFO] Worker exiting (pid: 1014)
[2021-06-28 16:46:24 +0200] [720] [INFO] Shutting down: Master
^C //.... After a few seconds I hit ctrl+c as nothing more was printed out
root@fullnode:/chia-blockchain#

**Comment by guydavis** [comment]
Good day! I think we now see the issue. You need to set your `-e plots_dir=/plots/folder1:/plots/folder2:/plots/folder3` when you have multiple folders of plots.  See this [FAQ entry](https://github.com/guydavis/machinaris/wiki/FAQ#farming).  Hope this helps!

**Reply by elexx** [reply]
I use only a single folder as destination and archive the plots with
```
archiving:
  target: local_rsync
  env:
    site_root: /plots
```
This has the advantage, that a new plot is created while the old one is being archived=moved to a slower HDD. Otherwise I loose a good portion of time while my plotter is moving the plot instead of plotting. I found this to be the most time efficient way (https://github.com/guydavis/machinaris/discussions/123#discussioncomment-933282)

**Comment by elexx** [comment, marked as answer]
Btw, my setup looks like this:
```
directories:
  tmp:
    - /plotting/tmp
  tmp2: /plotting_ram
  dst:
    - /plotting/dst

archiving:
  target: local_rsync
  env:
    site_root: /plots
```
I'm plotting with madmax. /plotting/tmp is a volume located on a SSD, the same goes with /plotting/dst. /plots is the common folder where my HDDs are mounted:
```
root@gold:/chia-blockchain# ls -la /plots/
total 76
drwxr-x---+ 12 root root   240 Jun 24 20:31 .
drwxr-xr-x   1 root root  4096 Jun 25 13:59 ..
drwxr-xr-x   3 1000 1000 12288 Jun 26 09:27 19c0dbe2-30e4-4734-92d9-2a81b049111b
drwxr-xr-x   3 1000 1000 12288 Jun 26 22:50 1ae119fa-294e-4449-b28f-a27141b9a77c
drwxr-xr-x   3 1000 1000 12288 Jun 27 21:10 3ec8901d-d783-48ec-834c-b7a9e84f2f39
[....]
```

**Reply by guydavis** [reply]
Yes, let me know if you have trouble creating the expected environment variable to add to your Docker run command.  

Example provided:
```
-e plots_dir=/plots/folder1:/plots/folder2:/plots/folder3
```

Please note this is a colon-separated list of in-container paths.  So `/plots` in-container, looks to map to this path on your Host OS: `/run/media/$(whoami)`  So, to be clear, list the folder names of `/run/media/$(whoami)` and add them above keeping `/plots/` in-container parent directory.  

Every folder listed above should contain *.plot files.  The folders passed to `plots_dir` should *not* be parent or higher folders which do *not* contain *.plot files.

Then restart the Machinaris container and browse to the Farming page.  For more details, see the [FAQ page](https://github.com/guydavis/machinaris/wiki/FAQ#farming). Hope this helps.

**Reply by elexx** [reply]
I see. I expected the page to display the plots as configured by
```
/root/.chia/mainnet/config/config.yaml
harvester:
  plot_directories:
  - /plots/19c0dbe2-30e4-4734-92d9-2a81b049111b
  - /plots/1ae119fa-294e-4449-b28f-a27141b9a77c
  - /plots/3ec8901d-d783-48ec-834c-b7a9e84f2f39
  - ...
```

**Reply by guydavis** [reply]
Yes, it looks like you added those yourself into the config.yaml for Chia services, or invoked `chia plots add -d` yourself from the CLI?  

Machinaris inherited the `plots_dir` environment variable from the official Chia docker which:

1. Invokes `chia plots add -d` for each directory provided in `plots_dir`
2. Which automtically adds the directory to the Chia config.yaml
3. Which automatically includes the directory when listing plots in the WebUI.

Hope this helps.  If so, please close this discussion by selecting 'Mark as Answer' on one of the comments.

**Reply by elexx** [reply]
You are absolutely right! I somehow totally missed this environment variable and edited the file myself. I just recreated the container with an appropriate environment variable and the plots are now visible.
